_episodes/09-cmorization.md

```diff
@@ -0,0 +1,1053 @@
+variable "gpp". The original units are therefore listed in the "global
+attributes" section as ``invalid_units``. The ESMValTool recognized that the
+units given in the original file did not match the units given in the CMOR
+table and therefore put the information about the units in the metadata.
+
+If we test this NetCDF file now with our CMOR checker
+```bash
+esmvaltool run recipe_check_fluxcom.yml --log_level debug
+```
+we will encounter an error again. The dataset is not correctly CMORized, and
+the first thing that the ESMValTool complains about is:
+```bash
+iris.exceptions.UnitConversionError: Cannot convert from unknown units. The
+"units" attribute may be set directly.
+```
+
+Ok, so let's fix the units of the "GPP" variable in the CMORizer. For that
+we add the following three lines to the code in the section
+``_extract_variable``:
+
+```python
+# convert data from gc/m2/day to kg/m2/s
+cube = cube / (1000 * 86400)
+cube.units = 'kg m-2 s-1'
+```
+
+The whole section should not look then look like this:
+
+```python
+def _extract_variable(cmor_info, attrs, filepath, out_dir):
+    """Extract variable."""
+    var = cmor_info.short_name
+    logger.info("Var is %s", var)
+    cubes = iris.load(filepath)
+    for cube in cubes:
+        # convert data from gc/m2/day to kg/m2/s
+        cube = cube / (1000 * 86400)
+        cube.units = 'kg m-2 s-1'
+
+        logger.info("Saving file")
+        utils.save_variable(cube,
+                            var,
+                            out_dir,
+                            attrs,
+                            unlimited_dimensions=['time'])
+
+```
+
+If we run the CMORizer script now again with the ``cmorize_obs`` call we
+get a NetCDF file that has fixed units, but has an "unknown" variable name.
+
+```bash
+...
+variables:
+        float unknown(time, lat, lon) ;
+                unknown:_FillValue = 1.e+20f ;
+                unknown:units = "kg m-2 s-1" ;
+        double time(time) ;
+                time:axis = "T" ;
+                time:units = "days since 1582-10-15 00:00:00" ;
+                time:standard_name = "time" ;
+                time:calendar = "gregorian" ;
+        double lat(lat) ;
+        double lon(lon) ;
+...
+```
+
+We will have to do some more code tweaking then. But we also have not fixed
+the problem with the coordinates ``lat`` and ``lon`` yet. There is no "units"
+or "standard_name" given for either of these coordinates which will cause a
+problem for the ESMValTool. Such some smaller formatting problems can occur
+relatively often for coordinates like ``lat``, ``lon`` or ``time``. This means
+that these problems need fixing in many CMORizers. Therefore there are common
+functions available within the ESMValTool that one can import and use in the
+new CMORizer script. The functions, written in python, are stored in the folder
+[utilities.py](https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/cmorizers/obs/utilities.py)
```

Comment:
The script utilities.py has already been introduced above.

Reply:
Restructured so this is different now